### DISCUSSION TITLE
Shapely for 3-dimension and more

### DISCUSSION THREAD

**Discussion by mvnayagam** [shapely/shapely#1353] [General]
Dear Developers and Users,

currently, I am using the Shapely package for my project in 2-dimensional cases.  In our project, we deal with Polygons( 2-dimensional) and polyhedrons (3-dimensional). Later we plan to apply our method for further higher dimensions, like 4-d, 5-d, etc.

With Shapely performance in the 2d case, we are excited to use it for further dimensions. But Shapely on the backend uses GEOS, which is for 2d polygons. so I hope that I can not use Shapely on any dimension higher than 2. 
Is it correct ?? (correct me if I am wrong ) 

is there any way to use Shapely on higher dimensional Polygon ( i.e on Polyhedra )?

I highly appreciate any suggestions on this topic!

thank you

---
Best regards
Muthu Vallinayagam,
Researcher,
TU Freiberg, Germany

**Comment by jorisvandenbossche** [comment]
Shapely is indeed wrapping GEOS, and so it is limited to the functionality of GEOS on this aspect. In general, GEOS supports storing 3D geometries, but the actual support for this third dimension in operations (eg overlay operations or spatial predicates) is quite limited (AFAIK quite some of those will ignore the third dimension). 

While improved 3D support could maybe be a possible topic of discussion in the GEOS project, I don't think it is likely that there will ever be support for higher dimensions like 4d, 5d, etc. 

For alternatives, you might want to take a look at CGAL (https://www.cgal.org/), although I have no experience with it (their docs seems to mention some things about nD, eg https://doc.cgal.org/latest/Kernel_d/index.html#Chapter_dD_Geometry_Kernel). There is a python package that provides basic bindings: https://github.com/scikit-geometry/scikit-geometry (although I think it only supports 2D and 3D geometries at the moment).

**Reply by mvnayagam** [reply]
Dear Joris,
thank you for your informative reply. 
3D support from shapely definitely will help many people in the physics stream.  I hope to see this functionality in the upcoming versions.
at present both 2D and 3D are vital for our project. (thank you for your suggestion, CGAL & scikit, I will give it a try. but do not think that it would be as handy as shapely. )

I am mainly using Polygon, Mutlipolygon, unary_union, and intersection. If I want to make a 3D version, ***what are the GEOS libs should I change?***

 the following tells how I work in 2D case,

*1. from the point list of series1 create Polygons (many polygons)
2. convert Polygons from step 1 to Multipolgon -> this represents series1
3. Repeat steps 1 and 2 for all series.  ( finally, if 10 series are available then there will be 10 Multipolygon)
4. find the intersection between Multipolygons from step 3*


I request your comments on the above points 

thank you

**Reply by jorisvandenbossche** [reply]
For reference, this was also asked on the geos-devel mailing list, see https://lists.osgeo.org/pipermail/geos-devel/2022-March/010674.html

I am personally not familiar enough with GEOS to be able to answer those questions. But I think the answer of Paul on the mailing list is also clear that an actual volumetric intersection in 3D  is not directly on the roadmap for GEOS.
